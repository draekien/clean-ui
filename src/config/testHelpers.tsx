import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
