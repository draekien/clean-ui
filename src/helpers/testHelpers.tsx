import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
