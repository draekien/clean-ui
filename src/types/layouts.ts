export type HorizontalAlignment = 'left' | 'center' | 'right';
export type VerticalAlignment = 'top' | 'center' | 'bottom';
export type Direction = 'row' | 'column';
export type Size = 'small' | 'medium' | 'large';
