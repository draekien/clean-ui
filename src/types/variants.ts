export type StatusVariant = 'success' | 'info' | 'error' | 'warning';
