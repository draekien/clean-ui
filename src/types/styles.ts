import { Theme } from '@theme-ui/css';
import { colors } from '../components/theme/theme.colors';
import { layouts } from '../components/theme/theme.layouts';

/**
 * The colors defined in the theme.
 * @remarks
 * See {@link colors} for more details.
 */
export type Color = keyof typeof colors;
/** The color computed from the theme. */
export type ComputedColor = (t: Theme) => string;
/**
 * The spacings defined in the theme.
 * @remarks
 * See {@link layouts.space} for more details.
 */
export type Spacing = keyof typeof layouts.space;
/**
 * The radii defined in the theme.
 * @remarks
 * See {@link layouts.radii} for more details.
 */
export type Radii = keyof typeof layouts.radii;
/**
 * The shadows defined in the theme.
 * @remarks
 * See {@link layouts.shadows} for more details.
 */
export type Shadows = keyof typeof layouts.shadows;
/**
 * The zIndices defined in the theme.
 * @remarks
 * See {@link layouts.zIndices} for more details.
 */
export type ZIndex = keyof typeof layouts.zIndices;
