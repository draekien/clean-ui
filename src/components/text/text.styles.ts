import { ThemeUIStyleObject } from 'theme-ui';
import { HorizontalAlignment } from '../../types/layouts';
import { colors } from '../theme/theme.colors';
import { textVariants } from '../theme/theme.variants';

export interface textCssProps {
  variant: keyof typeof textVariants;
  color: keyof typeof colors | 'inherit' | 'initial';
  textAlign: HorizontalAlignment;
  fullWidth: boolean;
  truncate: boolean;
}

export const textCss = ({
  variant,
  color,
  textAlign,
  truncate,
  fullWidth,
}: textCssProps): ThemeUIStyleObject => {
  const css: any = {
    variant: `text.${variant}`,
    color: color,
    display: fullWidth ? 'block' : 'inline-block',
    textAlign: textAlign,
    transition: 'all 300ms',
  };

  if (truncate) {
    css.overflow = 'hidden';
    css.textOverflow = 'ellipsis';
    css.whiteSpace = 'nowrap';
    css.display = 'inline-block';
    css.maxWidth = '100%';
  }

  return css;
};
