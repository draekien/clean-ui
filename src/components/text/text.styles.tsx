import { ThemeUIStyleObject } from 'theme-ui';
import { colors } from '../theme/theme.colors';
import { TextVariant } from './text';

export interface textCssProps {
  variant: TextVariant;
  color: keyof typeof colors;
}

export const textCss = ({ variant, color }: textCssProps): ThemeUIStyleObject => {
  const css: any = {
    variant: `text.${variant}`,
    color: color,
  };

  return css;
};
