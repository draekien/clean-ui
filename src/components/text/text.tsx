/** @jsxImportSource theme-ui */
import React from 'react';
import { Text as ThemeUiText } from 'theme-ui';
import { colors } from '../theme/theme.colors';
import { textVariants } from '../theme/theme.variants';
import { HorizontalAlignment } from '../../types/alignments';
import * as styles from './text.styles';

export interface TextProps {
  /** The type of DOM element to render
   * @default 'div'
   */
  as?: React.ElementType<any>;
  /** The text variant
   * @default 'body'
   */
  variant?: keyof typeof textVariants;
  /** The color of the text
   * @default 'text'
   */
  color?: keyof typeof colors;
  /** The text to render */
  children: React.ReactNode;
  /** horizontal alignment of the text
   * @default 'left'
   */
  textAlign?: HorizontalAlignment;
  /** should text take up the full width of the parent container */
  fullWidth?: boolean;
  /** should text be truncated instead of overflow */
  truncate?: boolean;
}

export const Text: React.FC<TextProps> = ({
  as,
  children,
  variant = 'body',
  color = 'text',
  textAlign = 'left',
  truncate = false,
  fullWidth = false,
}) => {
  return (
    <ThemeUiText
      as={as}
      sx={styles.textCss({ variant, color, textAlign, truncate, fullWidth })}>
      {children}
    </ThemeUiText>
  );
};
