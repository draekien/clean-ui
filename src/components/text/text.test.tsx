import React from 'react';
import { Text } from './text';
import { themedRender } from '../../helpers/testHelpers';
import { HorizontalAlignment } from '../../types/alignments';

describe('Text Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(<Text>Hero</Text>);

    expect(container).toMatchSnapshot();
  });

  it('Should render as block when fullwidth prop is true', () => {
    const { container } = themedRender(<Text fullWidth>Text</Text>);

    const styles = window.getComputedStyle(container.firstChild as Element);
    expect(styles.display).toBe('block');
  });

  it('Should render as inline-block when fullwidth prop is false', () => {
    const { container } = themedRender(<Text data-testid="text">Text</Text>);

    const styles = window.getComputedStyle(container.firstChild as Element);
    expect(styles.display).toBe('inline-block');
  });

  it('Should render with Hero font styles when variant prop is hero', () => {
    const { container } = themedRender(<Text variant="hero">Hero</Text>);

    const styles = window.getComputedStyle(container.firstChild as Element);
    expect(styles.fontSize).toBe('6rem');
    expect(styles.fontWeight).toBe('700');
  });

  it('Should render with Primary color when color prop is primary', () => {
    const { container } = themedRender(<Text color="primary">Primary Color</Text>);

    const styles = window.getComputedStyle(container.firstChild as Element);
    expect(styles.color).toBe('rgb(239, 128, 47)');
  });

  it('Should truncate text when the truncate flag is true', () => {
    const { container } = themedRender(
      <Text truncate>Some really long text that requires truncating</Text>
    );

    const styles = window.getComputedStyle(container.firstChild as Element);
    expect(styles.overflow).toBe('hidden');
    expect(styles.textOverflow).toBe('ellipsis');
    expect(styles.whiteSpace).toBe('nowrap');
    expect(styles.maxWidth).toBe('100%');
  });

  const alignments: HorizontalAlignment[] = ['left', 'center', 'right'];

  alignments.map((textAlign) => {
    it(`Should align text ${textAlign} when textAlign is ${textAlign}`, () => {
      const { container } = themedRender(<Text textAlign={textAlign}>Text</Text>);

      const styles = window.getComputedStyle(container.firstChild as Element);
      expect(styles.textAlign).toBe(textAlign);
    });
  });
});
