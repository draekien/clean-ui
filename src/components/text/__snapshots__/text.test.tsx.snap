// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Should match snapshot 1`] = `
<div>
  <div
    class="css-zfdk4x-Text css-vurnku"
  >
    Hero
  </div>
</div>
`;
